### PR TITLE
Don't rebuild stack if it is up-to-date

### PR DESCRIPTION
**Pull request by bixuanzju**
Ref #811

I've tested on my local fork, it works. But I am not sure if I am doing it in a robust way. Perhaps I could have made a customised `readProcess` just like what you have as [`callProcess`](https://github.com/commercialhaskell/stack/blob/master/src/System/Process/Run.hs#L63).

### CODE REVIEW

File: src/Stack/Upgrade.hs [hunk 1]
```diff
@@ -38,6 +41,9 @@ upgrade gitRepo mresolver = withSystemTempDirectory "stack-upgrade" $ \tmp' -> d
     tmp <- parseAbsDir tmp'
     mdir <- case gitRepo of
       Just repo -> do
+        remote <- liftIO $ readProcess "git" ["ls-remote", repo, "master"] []
+        let latestCommit = head . words $ remote
+        when (latestCommit == $(gitHash)) $ error "Already up-to-date, no upgrade required"
```
Comment (snoyberg) [inline src/Stack/Upgrade.hs]: Just for future reference: you can replace `$(gitHash)` with `$gitHash`.

Comment (snoyberg) [inline src/Stack/Upgrade.hs]: Also: I don't think we should consider this an error condition, as that could break automated tooling that would simply want to know that it's using the newest stack. Can you instead just print a message and return `Nothing` like the non-Git case does?

Comment (bixuanzju) [inline src/Stack/Upgrade.hs]: Sure!

Comment (bixuanzju) [inline src/Stack/Upgrade.hs]: @snoyberg Done! Please have a look

**Comment by snoyberg** [comment]
Looks great, thanks! I'll just wait for the Travis green light and then merge.

**Comment by snoyberg** [comment]
Thanks!

**Pull request merged**